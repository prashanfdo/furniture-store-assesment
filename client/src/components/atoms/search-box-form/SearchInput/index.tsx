import { MagnifyingGlassIcon } from "@heroicons/react/20/solid";
import React, { ForwardedRef, forwardRef } from "react";
import { UseFormRegisterReturn } from "react-hook-form";

type SearchInputProps = UseFormRegisterReturn;

const SearchInput = forwardRef((props: SearchInputProps, ref: ForwardedRef<HTMLInputElement>) => {
  const { ...registerProps } = props;
  return (
    <>
      <input
        className="w-full px-3 py-2 bg-gray-200 border-transparent border-none rounded-tl-lg rounded-bl-lg outline-transparent"
        type="text"
        {...registerProps}
        ref={ref}
        autoComplete="off"
      />
      <button className="grid w-10 text-white transition-colors border-transparent rounded-tr-lg rounded-br-lg place-items-center bg-coral-500 hover:bg-coral-300">
        <MagnifyingGlassIcon className="w-5 h-5 font-bold" />
      </button>
    </>
  );
});

export default SearchInput;
