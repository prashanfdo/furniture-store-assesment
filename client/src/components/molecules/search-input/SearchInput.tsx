import styled from "@emotion/styled";
import React, { useEffect, useState } from "react";
import { InputWithButton } from "./InputWithButton";
import { SuggestionsList } from "./SuggestionsList";
import useSearchForm from "./useSearchForm";

const FormContainer = styled.form`
  display: grid;
  grid-template-columns: 1fr auto;
  width: 500px;
  position: relative;
`;

type SearchInputProps = Parameters<typeof useSearchForm>[0] &
  React.ComponentProps<typeof SuggestionsList>;

const SearchInput: React.FC<SearchInputProps> = ({
  initialSearchText,
  onSearchSubmit,
  onSearchChange,
  suggestions,
}: SearchInputProps) => {
  const searchForm = useSearchForm({ initialSearchText, onSearchSubmit, onSearchChange });

  return (
    <>
      <FormContainer onSubmit={searchForm.onSubmit}>
        <InputWithButton register={searchForm.register} />
        <SuggestionsList suggestions={suggestions} />
      </FormContainer>
    </>
  );
};

export default SearchInput;
