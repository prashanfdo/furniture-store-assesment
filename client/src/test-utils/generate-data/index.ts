export * from "./article";
export * from "./category";
export * from "./search";
