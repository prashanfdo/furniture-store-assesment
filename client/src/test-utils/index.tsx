import { render, RenderOptions, renderHook, RenderHookOptions, RenderHookResult } from "@testing-library/react";
import fs from "fs";
import React, { FC, ReactElement } from "react";

import user from "@testing-library/user-event";
import { ProvidersForTests } from "providers";

type GraphQlMocks = React.ComponentProps<typeof ProvidersForTests>["mocks"];

type WrapperProps = React.ComponentProps<typeof ProvidersForTests> & {
  children: React.ReactNode;
};
const Wrapper: FC<WrapperProps> = ({ children, mocks, initialRoute }: WrapperProps) => {
  return (
    <ProvidersForTests mocks={mocks} initialRoute={initialRoute}>
      {children}
    </ProvidersForTests>
  );
};
const wrapper =
  (mocks: GraphQlMocks, initialRoute?: string) =>
  ({ children }: { children: React.ReactNode }) =>
    (
      <Wrapper mocks={mocks} initialRoute={initialRoute}>
        {children}
      </Wrapper>
    );

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper"> & { gqlMocks: GraphQlMocks; initialRoute: string },
) => {
  const view = render(ui, { wrapper: wrapper(options?.gqlMocks, options?.initialRoute), ...options });
  const style = document.createElement("style");
  style.innerHTML = fs.readFileSync("src/index.out.css", "utf8");
  document.head.appendChild(style);

  return view;
};

export const customRenderHook = <T, R>(
  callback: (props: T) => R,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { gqlMocks?: GraphQlMocks; initialRoute?: string },
): RenderHookResult<R, T> => {
  return renderHook(callback, { ...options, wrapper: wrapper(options?.gqlMocks, options?.initialRoute) });
};

export { faker } from "@faker-js/faker";
export * from "@testing-library/react";
export * from "./generate-data";
export { customRender as render };
export { customRenderHook as renderHook };
export { user };
