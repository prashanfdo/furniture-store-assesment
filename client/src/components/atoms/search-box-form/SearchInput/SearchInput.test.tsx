import user from "@testing-library/user-event";
import { render, screen } from "test-utils";
import SearchInput from "./index";

describe("SearchInput", () => {
  beforeEach(() => {
    user.setup();
  });

  it("should render", () => {
    const handleChange = jest.fn();
    const handleBlur = jest.fn();
    const { container } = render(<SearchInput name="asd" onChange={handleChange} onBlur={handleBlur} />);
    expect(container).toMatchSnapshot();
  });

  it("should render textbox and button", () => {
    render(<SearchInput name="search" onChange={jest.fn()} onBlur={jest.fn()} />);
    expect(screen.getByRole("textbox")).toBeInTheDocument();
    expect(screen.getByRole("button")).toBeInTheDocument();
  });

  it("should invoke onchange callback", async () => {
    const handleChange = jest.fn();
    const handleBlur = jest.fn();
    render(<SearchInput name="search" onChange={handleChange} onBlur={handleBlur} />);
    const textbox = screen.getByRole<HTMLInputElement>("textbox");
    await user.type(textbox, "test");
    expect(textbox).toHaveValue("test");
    // TODO: Fix this test
    // await waitFor(() => expect(handleChange).toHaveBeenCalledTimes("test".length));
    // await user.click(screen.getByRole("button"));
  });
});
