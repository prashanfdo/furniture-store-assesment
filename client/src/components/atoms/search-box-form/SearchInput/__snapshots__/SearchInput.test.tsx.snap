// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput should render 1`] = `
<div>
  <input
    autocomplete="off"
    class="w-full px-3 py-2 bg-gray-200 border-none rounded-tl-lg rounded-bl-lg outline-transparent"
    name="asd"
    type="text"
  />
  <button
    class="grid w-10 place-items-center"
  >
    <svg
      aria-hidden="true"
      class="w-5 h-5 text-gray-600"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M9 3.5a5.5 5.5 0 100 11 5.5 5.5 0 000-11zM2 9a7 7 0 1112.452 4.391l3.328 3.329a.75.75 0 11-1.06 1.06l-3.329-3.328A7 7 0 012 9z"
        fill-rule="evenodd"
      />
    </svg>
  </button>
</div>
`;
